pkg/manifests/external_dns.go

```diff
@@ -0,0 +1,213 @@
+			},
+		},
+	}
+}
+
+func newExternalDNSClusterRoleBinding(conf *config.Config, dnsConfig *ExternalDnsConfig) *rbacv1.ClusterRoleBinding {
+	return &rbacv1.ClusterRoleBinding{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ClusterRoleBinding",
+			APIVersion: "rbac.authorization.k8s.io/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   dnsConfig.ResourceName,
+			Labels: topLevelLabels,
+		},
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: "rbac.authorization.k8s.io",
+			Kind:     "ClusterRole",
+			Name:     dnsConfig.ResourceName,
+		},
+		Subjects: []rbacv1.Subject{{
+			Kind:      "ServiceAccount",
+			Name:      dnsConfig.ResourceName,
+			Namespace: conf.NS,
+		}},
+	}
+}
+
+func newExternalDNSConfigMap(conf *config.Config, externalDnsConfig *ExternalDnsConfig) (*corev1.ConfigMap, string) {
+	js, err := json.Marshal(&map[string]interface{}{
+		"tenantId":                    externalDnsConfig.TenantId,
+		"subscriptionId":              externalDnsConfig.Subscription,
+		"resourceGroup":               externalDnsConfig.ResourceGroup,
+		"userAssignedIdentityID":      conf.MSIClientID,
+		"useManagedIdentityExtension": true,
+		"cloud":                       conf.Cloud,
+		"location":                    conf.Location,
+	})
+	if err != nil {
+		panic(err)
+	}
+	hash := sha256.Sum256(js)
+	return &corev1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ConfigMap",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      externalDnsConfig.ResourceName,
+			Namespace: conf.NS,
+			Labels:    topLevelLabels,
+		},
+		Data: map[string]string{
+			"azure.json": string(js),
+		},
+	}, hex.EncodeToString(hash[:])
+}
+
+func newExternalDNSDeployment(conf *config.Config, configMapHash string, externalDnsConfig *ExternalDnsConfig) *appsv1.Deployment {
+	replicas := int32(1)
+	return &appsv1.Deployment{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Deployment",
+			APIVersion: "apps/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      externalDnsConfig.ResourceName,
+			Namespace: conf.NS,
+			Labels:    topLevelLabels,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas:             &replicas,
+			RevisionHistoryLimit: util.Int32Ptr(2),
+			Selector:             &metav1.LabelSelector{MatchLabels: map[string]string{"app": externalDnsConfig.ResourceName}},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app":                externalDnsConfig.ResourceName,
+						"checksum/configmap": configMapHash[:16],
+					},
+				},
+				Spec: *WithPreferSystemNodes(&corev1.PodSpec{
+					ServiceAccountName: externalDnsConfig.ResourceName,
+					Containers: []corev1.Container{*withLivenessProbeMatchingReadiness(withTypicalReadinessProbe(7979, &corev1.Container{
+						Name:  "controller",
+						Image: path.Join(conf.Registry, "/oss/kubernetes/external-dns:v0.11.0.2"),
+						Args: []string{
+							"--provider=azure",
```

Comment:
wanted to change these into consts but didn't since you mentioned reorganizing these (which is awesome btw, this is SO much easier to read), but could you put stuff like azure and the "ingress" source as constants now that the placement of these strings is more stable?

Reply:
what is the value of making these consts? I don't see the value of making "azure" a const because it is extremely unlikely that will ever change, provider is only referenced in this single place, and what it means is most clear leaving it where it's used.

Source possibly makes more sense to move to a const since it's reasonable that we'd add new ones in the future (gateway potentially). However, the usage for source is that you pass each one in with a separate `--source` flag (https://github.com/kubernetes-sigs/external-dns/blob/master/pkg/apis/externaldns/types.go#L406). That would mean your const would actually be an array and you'd have to add logic here handling that making it more difficult to read than the nicely formatted string we see below. I also don't see any advantages to doing this for the same reasons as the azure const so it just would hurt code readability with no real benefits.

Reply:
rather than for future-proofing for change, I think it's just easier in case that string ends up being used elsewhere, like in other resources we end up deploying. 

Just a thought, cool to just resolve this as well if you feel necessary